primitives/xcm/src/barriers.rs

```diff
@@ -77,3 +77,97 @@ impl<T: Contains<MultiLocation>> ShouldExecute for AllowTopLevelPaidExecutionDes
 		}
 	}
 }
+
+/// Make sure that not withdrawable assets are handle properly according to the XCM design:
+/// - forbid any ReserveAssetDeposited instruction that contains not withdrawable asset(s) as
+/// parameter.
+/// - Morph some specific messages patterns that try to Widraw then Deposit not withdrawable
+/// asset(s).
+pub struct NotWithdrawableAssetsBarrier<IsWithdrawable, InnerBarrier>(
+	core::marker::PhantomData<(IsWithdrawable, InnerBarrier)>,
+);
+
+impl<IsWithdrawable: crate::IsWithdrawable, InnerBarrier: ShouldExecute> ShouldExecute
+	for NotWithdrawableAssetsBarrier<IsWithdrawable, InnerBarrier>
+{
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		instructions: &mut Xcm<Call>,
+		max_weight: Weight,
+		weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		InnerBarrier::should_execute(origin, instructions, max_weight, weight_credit)?;
+		let maybe_morphed_message = match &instructions.0[..] {
+			&[WithdrawAsset(ref assets), ClearOrigin, BuyExecution {
```

Comment:
@gavofyork or @KiChjang, what do you think of this implementation? It seems that we can't really do better than mutating the XCM fragment to change its instruction if we want to stay compliant with XCM ideas. However I'm worried this could lead to issues (like weight issues) or additional risks.
What is your opinion ?

Reply:
I responded here: https://forum.polkadot.network/t/how-to-handle-not-withdrawable-assets-properly-with-xcm/2270/2